kratos/utilities/auxilary_global_master_slave_relation.h

```diff
@@ -0,0 +1,226 @@
+     * @param SlaveEquationId the slave equation id for which this class is being constructed.
+     */
+    AuxilaryGlobalMasterSlaveRelation(IndexType SlaveEquationId = 0) : IndexedObject(SlaveEquationId), mConstant(0.0)
+    {
+    }
+
+    /**
+     * @brief Function to get the slave equation Id corresponding to this constraint.
+     * @param Constant the value of the constant to be assigned.
+     */
+    IndexType SlaveEquationId() const { return this->Id(); }
+
+    /**
+     * @brief Function to set the lefthand side of the constraint (the slave dof value)
+     * @param LhsValue the value of the lhs (the slave dof value)
+     */
+    void SetLHSValue(const double& LhsValue)
+    {
+        mLockObject.SetLock();
+            mLhsValue = LhsValue;
+        mLockObject.UnSetLock();
+    }
+
+    /**
+     * @brief Function to update the righthand side of the constraint (the combination of all the master dof values and constants)
+     * @param RHSValue the value of the lhs (the slave dof value)
+     */
+    void SetRHSValue(const double& RhsValue)
+    {
+        mRhsValue = RhsValue;
+    }
+    void UpdateRHSValue(const double RhsValueUpdate)
+    {
+        mLockObject.SetLock();
+            mRhsValue = mRhsValue + RhsValueUpdate;
+        mLockObject.UnSetLock();
+    }
+
+    // Get number of masters for this slave
+    IndexType GetNumberOfMasters() const
+    {
+        return mMasterEquationIdVector.size();
+    }
+
+    /**
+     * @brief this determines the master equation IDs connected to this constraint
+     * @param rResult the elemental equation ID vector
+     */
+    virtual void EquationIdVector(IndexType& rSlaveEquationId,
+                                  EquationIdVectorType& rMasterEquationIds)
+    {
+        if (rMasterEquationIds.size() != 0 || rMasterEquationIds.size() == 0)
+            rMasterEquationIds.resize(this->GetNumberOfMasters(), false);
+
+        rSlaveEquationId = this->SlaveEquationId();
+
+        for (IndexType i = 0; i < this->GetNumberOfMasters(); ++i)
+        {
+            rMasterEquationIds[i] = mMasterEquationIdVector[i];
+        }
+    }
+
+    /**
+     * @brief   this is called during the assembling process in order
+     *          to calculate all elemental contributions to the global system
+     * matrix and the right hand side
+     * @param rMasterWeightsVector the elemental left hand side matrix
+     * @param rConstant the elemental right hand side
+     */
+    virtual void CalculateLocalSystem(VectorType &rMasterWeightsVector,
+                                      double &rConstant)
+    {
+        if (rMasterWeightsVector.size() != 0 || rMasterWeightsVector.size() == 0)
+        {
+            rMasterWeightsVector.resize(this->GetNumberOfMasters(), false);
+        }
+
+        for (IndexType i = 0; i < this->GetNumberOfMasters(); ++i)
+            rMasterWeightsVector(i) = mMasterWeightsVector[i];
+
+
+        mLockObject.SetLock();
```

Comment:
do you really need this lock? it is strange to protect an assignement operator as you do here...

I mean, if two threads act on this, the result will be nondeterministic even if you do the lock, depending on which thread executes it last.

i really think this is a bug

Reply:
As the mRhsValue and mLhsValue and always the same, even if multiple threads operate , the value of mConstant can have only one value. 
With this lock I am trying to avoid that multiple threads write (same value) to mConstant.